### PR TITLE
Issue_439: Convert Pytest-CVP Username No Password test case to Vane #439

### PR DESCRIPTION
**Pull request by dhruvgoel9**
# Please include a summary of the changes

Converted test_security_rp_username_password.py as per the vane style guide along with the updated following files.
test_definition.yaml: Updated test def with testcase NRFU5.8
test_definition.yaml.j2: Updated J2 with testcase NRFU5.8
nrfu_tests/test_security_rp_username_password.py: Added converted testsuite file for testcases NRFU5.8

# Any specific logic/part of code you need extra attention on

From the output of show running-config all section ^username, verified that the all username configured with a password , if usernames are not found then testcase will skip.

# Include the Issue number and link

#439 

# List/Attach any dependencies/past issues that are required for this change/provide context

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [ ] New feature (non-breaking change which adds functionality)
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [x] Other (NRFU test case)

# Effort required on reviewers end

- - [ ] Easy
- - [x] Medium
- - [ ] Hard 

# How Has This Been Tested?

Unit tests:

Tested test_security_rp_username_password with username have password configured: Test case passed
[pass_report.docx](https://github.com/aristanetworks/vane/files/12542207/pass_report.docx)

Tested test_security_rp_username_password with username have password configured using j2 test definition: Test case passed
[pass_report_j2.docx](https://github.com/aristanetworks/vane/files/12542209/pass_report_j2.docx)

Tested test_security_rp_username_password with username are not configured with password: Test case failed
[fail_report.docx](https://github.com/aristanetworks/vane/files/12542213/fail_report.docx)

Tested test_security_rp_username_password with usernames are not found (Used hardcoded output): Test case failed with assertion message
[assert_fail_no_username_found.docx](https://github.com/aristanetworks/vane/files/12542216/assert_fail_no_username_found.docx)

HTML files:
[html report.zip](https://github.com/aristanetworks/vane/files/12542218/html.report.zip)

### CODE REVIEW

File: nrfu_tests/test_security_rp_username_password.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2023 Arista Networks, Inc.  All rights reserved.
```
Comment (hsisalgslab) [inline nrfu_tests/test_security_rp_username_password.py:1]: add label enhancement

File: nrfu_tests/test_definition.yaml.j2 [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  {% set master_data = testcase_data %}
+  testcases:
+    - name: test_security_rp_username_password
+      description: Test case for verification of username password status.
+      test_id: NRFU5.8
+      show_cmd: show running-config all section ^username
+      expected_output: null
+      test_criteria: All usernames should be configured with a password.
+      report_style: modern
+      criteria: {{ global_dut_filter["criteria"] }}
```
Comment (hsisalgslab) [inline nrfu_tests/test_definition.yaml.j2:11]:
```suggestion
      criteria: {{ global_dut_filter.criteria }}
```
access like this

File: nrfu_tests/test_security_rp_username_password.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+from vane.logger import logger
+from vane.config import dut_objs, test_defs
+from vane import tests_tools
+
+TEST_SUITE = "nrfu_tests"
+
+
+@pytest.mark.nrfu_test
+@pytest.mark.security
+class SecurityRPUsernamePasswordTests:
```
Comment (hsisalgslab) [inline nrfu_tests/test_security_rp_username_password.py:19]:
```suggestion
class SecurityRPUsernameNoPasswordTests:
```
can we update like this ?

File: nrfu_tests/test_security_rp_username_password.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    """
+
+    dut_parameters = tests_tools.parametrize_duts(TEST_SUITE, test_defs, dut_objs)
+    test_duts = dut_parameters["test_security_rp_username_password"]["duts"]
+    test_ids = dut_parameters["test_security_rp_username_password"]["ids"]
+
+    @pytest.mark.parametrize("dut", test_duts, ids=test_ids)
+    def test_security_rp_username_password(self, dut, tests_definitions):
+        """
+        TD: Test case for verification of username password status.
```
Comment (hsisalgslab) [inline nrfu_tests/test_security_rp_username_password.py:31]:
```suggestion
        TD: Test case for verification of username and password status.
```
check for applicable cases

File: nrfu_tests/test_security_rp_username_password.py (hunk 1 above)
Comment (dhruvgoel9) [reply nrfu_tests/test_security_rp_username_password.py:1]: enhancement label is not required

File: nrfu_tests/test_security_rp_username_password.py (hunk 3 above)
Comment (dhruvgoel9) [reply nrfu_tests/test_security_rp_username_password.py]: class name is updated as per discussion with vishwajeet

**Review by hsisalgslab — APPROVED** [review]

File: nrfu_tests/test_definition.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+- name: nrfu_tests
+  testcases:
+  - name: test_security_rp_username_password
+    description: Test cases for verification of password configuration for user.
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_definition.yaml:4]:
```suggestion
    description: Test cases for verification of password configuration for user accounts.
```

File: nrfu_tests/test_definition.yaml.j2 [hunk 6]
```diff
@@ -0,0 +1,12 @@
+- name: nrfu_tests
+  {% set master_data = testcase_data %}
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_definition.yaml.j2:2]: is this required here ?

File: nrfu_tests/test_security_rp_username_password.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            dut(dict): details related to a particular DUT
+            tests_definitions(dict): test suite and test case parameters.
+        """
+        tops = tests_tools.TestOps(tests_definitions, TEST_SUITE, dut)
+        self.output = ""
+        tops.actual_output = {}
+        tops.expected_output = {}
+
+        # Forming output message if test result is passed
+        tops.output_msg = "All usernames are configured with a password."
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:42]:
```suggestion
        tops.output_msg = "All user accounts are configured with a password."
```

File: nrfu_tests/test_definition.yaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    test_id: NRFU5.8
+    show_cmd: show running-config all section ^username
+    expected_output: null
+    test_criteria: All usernames should be configured with a password. # Setting test case’s pass/fail criteria for reporting
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_definition.yaml:8]:
```suggestion
    test_criteria: All user accounts should be configured with a password. # Setting test case’s pass/fail criteria for reporting
```

File: nrfu_tests/test_security_rp_username_password.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    """
+
+    dut_parameters = tests_tools.parametrize_duts(TEST_SUITE, test_defs, dut_objs)
+    test_duts = dut_parameters["test_security_rp_username_password"]["duts"]
+    test_ids = dut_parameters["test_security_rp_username_password"]["ids"]
+
+    @pytest.mark.parametrize("dut", test_duts, ids=test_ids)
+    def test_security_rp_username_password(self, dut, tests_definitions):
+        """
+        TD: Test cases for verification of password configuration for user.
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:31]: Update as per comment on test def. Applicable at all other places.

File: nrfu_tests/test_security_rp_username_password.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            logger.info(
+                "On device %s, output of %s command is:\n%s\n",
+                tops.dut_name,
+                tops.show_cmd,
+                output,
+            )
+            self.output += (
+                f"\nOn device {tops.dut_name}, output of {tops.show_cmd} command is:\n{output}\n"
+            )
+            assert output, "Username details are not found in the output."
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:59]:
```suggestion
            assert output, "User account details are not found in the output."
```

File: nrfu_tests/test_security_rp_username_password.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+            usernames = output.split("\n")
+            for user in usernames:
+                if user:
+                    username = user.split()[1]
+                    password = user.split()[-1]
+                    tops.expected_output.update({username: {"username_password_found": True}})
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:66]:
```suggestion
                    tops.expected_output.update({username: {"password_configured": True}})
```

File: nrfu_tests/test_security_rp_username_password.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                    if password == "nopassword":
+                        tops.actual_output.update({username: {"username_password_found": False}})
+                    else:
+                        tops.actual_output.update({username: {"username_password_found": True}})
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:70]:
```suggestion
                    tops.actual_output.update({username: {"username_password_found": True}})
                    if password == "nopassword":
                        tops.actual_output.update({username: {"username_password_found": False}})
```

File: nrfu_tests/test_security_rp_username_password.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+                for username in tops.expected_output:
+                    actual_username_detail = tops.actual_output.get(username, {}).get(
+                        "username_password_found"
+                    )
+                    if not actual_username_detail:
+                        password_not_configured.append(username)
+                if password_not_configured:
+                    tops.output_msg += (
+                        "Following users are not configured with a"
+                        f" password:\n{', '.join(password_not_configured)}"
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:85]:
```suggestion
                        "Following user accounts are not configured with a"
                        f" password:{', '.join(password_not_configured)}"
```

**Review by vishwajeetgslab — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_definition.yaml [hunk 14]
```diff
@@ -0,0 +1,12 @@
+- name: nrfu_tests
+  testcases:
+    - name: test_security_rp_username_password
+      description: Test cases for verification of password configuration for user accounts.
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_definition.yaml:4]:
```suggestion
      description: Test case for verification of password configuration for user accounts.
```

File: nrfu_tests/test_security_rp_username_password.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2023 Arista Networks, Inc.  All rights reserved.
+# Arista Networks, Inc. Confidential and Proprietary.
+
+"""
+Test cases for verification of password configuration for user accounts.
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:5]:
```suggestion
Test case for verification of password configuration for user accounts.
```

File: nrfu_tests/test_security_rp_username_password.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+                for username in tops.expected_output:
+                    actual_username_detail = tops.actual_output.get(username, {}).get(
+                        "password_configured"
+                    )
+                    if not actual_username_detail:
+                        password_not_configured.append(username)
+                if password_not_configured:
+                    tops.output_msg += (
+                        "Following user accounts are not configured with a"
+                        f" password:\n{', '.join(password_not_configured)}"
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_security_rp_username_password.py:88]:
```suggestion
                        f" password: {', '.join(password_not_configured)}"
```

**Review by vishwajeetgslab — CHANGES_REQUESTED** [review]

**Review by vishwajeetgslab — APPROVED** [review]

**Review by kgrozis-arista — APPROVED** [review]

**Pull request merged**